### PR TITLE
Override `ContractCreationProcessor` for receiver-sig aware dispatch

### LINKED ISSUE #7387: Subclass `ContractCreationProcessor`
Since the modularized service will not unilaterally modify account balances, we need to subclass the Besu `ContractCreationProcessor` to replace the `decrementBalance(sender) -> incrementBalance(receiver)` pattern with conceptually a single `transferWithReceiverSigCheck(sender, receiver)` call.

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #7387 
 - Finishes override of `ContractCreationProcessor`, replacing direct `sender.decrementBalance(frame.getValue())` and `contract.incrementBalance(frame.getValue())` calls with a `HederaWorldUpdater#tryTransferFromContract(Address, Address, long, boolean)` invocation.

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java [hunk 1]
```diff
@@ -16,20 +16,81 @@
...
+
+    @Override
+    public void start(@NonNull final MessageFrame frame, @NonNull final OperationTracer tracer) {
+        final var addressToCreate = frame.getContractAddress();
+        final var contract =
+                frame.getWorldUpdater().getOrCreate(addressToCreate).getMutable();
+
+        if (alreadyCreated(contract)) {
+            halt(frame, COLLISION_HALT_REASON);
+            tracer.traceAccountCreationResult(frame, COLLISION_HALT_REASON);
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:71]: I understand the comment that "insufficient gas" is the accepted reason to return for "contract already exists" but it's a shame that the information gets lost.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java [hunk 2]
```diff
@@ -16,20 +16,81 @@
...
+        }
+    }
+
+    private void halt(final MessageFrame frame, @NonNull final Optional<ExceptionalHaltReason> reason) {
+        frame.setState(MessageFrame.State.EXCEPTIONAL_HALT);
+        frame.setExceptionalHaltReason(reason);
+    }
+
+    private boolean alreadyCreated(final MutableAccount account) {
+        return account.getNonce() > 0 || account.getCode().size() > 0;
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:94]: Are there contracts with code size == 0 that nevertheless have nonce > 0 (they've been called)?  Just curious.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/processors/CustomContractCreationProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        verify(frame).setState(MessageFrame.State.EXCEPTIONAL_HALT);
+        verify(tracer).traceAccountCreationResult(frame, maybeReasonToHalt);
+    }
+
+    @Test
+    void haltsWithInsufficientGasIfContractExistsWithNonEmptyCode() {
+        given(frame.getContractAddress()).willReturn(EIP_1014_ADDRESS);
+        given(frame.getWorldUpdater()).willReturn(worldUpdater);
+        given(contract.getMutable()).willReturn(mutableContract);
+        given(mutableContract.getCode()).willReturn(CONTRACT_CODE.getBytes());
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/processors/CustomContractCreationProcessorTest.java:144]: mockito question:  why doesn't this test need a getNonce() -> 0 given?  is this a permissive vs strict thing?

**Review by david-bakin-sl — APPROVED** [review]
LGTM

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:71]: True, it's interesting they chose `INSUFFICIENT_GAS`. 🤔

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:94]: The [initial contract nonce](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/ProcessorModule.java#L29) is `1`, so _if_ it's possible to create a contract whose bytecode is empty, then we would have the case you describe.

I'm not sure if it's possible to create a contract with empty bytecode tho. 🤔

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/processors/CustomContractCreationProcessorTest.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/processors/CustomContractCreationProcessorTest.java:144]: Mockito returns `0` from any mocked method that returns a numeric type; so you don't have to configure that explicitly.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:71]: Ouch.
I wonder if this is a place for us to break trend and provide more information for devs. Not now but something to keep track of.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java [hunk 4]
```diff
@@ -16,20 +16,81 @@
...
+        final var contract =
+                frame.getWorldUpdater().getOrCreate(addressToCreate).getMutable();
+
+        if (alreadyCreated(contract)) {
+            halt(frame, COLLISION_HALT_REASON);
+            tracer.traceAccountCreationResult(frame, COLLISION_HALT_REASON);
+        } else {
+            final var updater = proxyUpdaterFor(frame);
+            // A contract creation is never a delegate call, hence the false argument below
+            final var maybeReasonToHalt = updater.tryTransferFromContract(
```
Comment (Nana-EC) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:75]: Q: just checking does the no value scenario get covered here? That is does it work and not fail when no value was provided to transfer. Something to add to tests if not already covered.

**Review by Nana-EC — APPROVED** [review]
LG
One question and potential test suggestion if it's not already internally handled

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java (hunk 4 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java:75]: I'm thinking this method should just immediately return `Optional.empty()` if asked to transfer zero value.

Will add that to its javadoc in my next PR.

**Pull request merged**